File: javapatterns/src/main/java/org/iut/nantes/state/canards/etats/SurLeau.java
package org.iut.nantes.state.canards.etats;

import org.iut.nantes.state.canards.Canard;

public class SurLeau implements Etat{

    private Canard canard;

    public SurLeau (Canard canard) {
        this.canard = canard;
    }

    @Override
    public void avancer() {
        System.out.println("Je nage sur l'eau.");
    }

    @Override
    public void plongeon() {
        canard.changeEtatSousLeau();
    }

    @Override
    public void trempette() {

    }

    @Override
    public void decollage() {
        canard.changeEtatDansLair();
    }

    @Override
    public void palmeATerre() {
        canard.changeEtatSurTerre();
    }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/state/canards/etats/SurLeau.kt
package org.iut.nantes.state.canards.etats

import org.iut.nantes.state.canards.Canard
import org.iut.nantes.state.canards.etats.Etat

class SurLeau(private val canard: Canard)  : Etat {

    override fun avancer() {
        println("Je nage")
    }
    override fun plongeon() {
        canard.changeEtatDansEau()
    }

    override fun trempette() {
        println("Je nage déjà")
    }

    override fun decollage() {
        canard.changeEtatDansAir()
    }

    override fun palmeATerre() {
        canard.changeEtatSurTerre()
    }
}